### ISSUE TITLE
Lack of support for pooling layer parameter channels_first

### ISSUE THREAD

**Issue by hunyadi** [Dobiasd/frugally-deep#121]
We have a model that averages contributions of a particular feature across time steps with a layer of
```
GlobalAveragePooling1D(data_format="channels_first")
```

However, when we execute `convert_model.py` to convert the trained HD5 model into the JSON that _frugally-deep_ takes, we run into an assertion error:
```
Traceback (most recent call last):
 File “convert_model.py”, line 665, in <module>
   main()
 File “convert_model.py”, line 661, in main
   convert(in_path, out_path)
 File “convert_model.py”, line 641, in convert
   json_output[‘trainable_params’] = get_all_weights(model)
 File “convert_model.py”, line 461, in get_all_weights
   result = merge_two_disjunct_dicts(result, get_all_weights(layer))
 File “convert_model.py”, line 458, in get_all_weights
   assert layer.data_format == ‘channels_last’
AssertionError
```

It looks like _frugally-deep_ does not support `channels_first`. In our model, a workaround is to use two consecutive layers in a sequential model:
```
model.add(Permute((2,1)))
model.add(GlobalAveragePooling1D(data_format="channels_last"))
```
but it looks more convoluted and is definitely harder to understand the intent.

The model (both with the original pooling layer and the workaround) trains and predicts as expected in Python.

**Comment by Dobiasd** [comment]
Oh, interesting. I guess this concerns all implemented pooling layers, i.e., `AveragePooling1D`, `AveragePooling2D`, `MaxPooling1D`, `MaxPooling2D`, `GlobalAveragePooling1D`, `GlobalAveragePooling2D`, `GlobalMaxPooling1D` and `GlobalMaxPooling2D`.

So is your global Keras setting `channels_first`, or is your global setting `channels_last` and only for the pooling you want to use `channels_first`?

**Comment by hunyadi** [comment]
We did not make extensive tests with all parameters (all of default, `channels_last`, `channels_first` with various other pooling layers), it was one of our engineers who reported the assertion error with `channels_first` but then when we implemented the workaround with `channels_last` and the extra `Permute` layer, all worked fine. We are using a stock installation of Keras/TensorFlow, which has `channels_last` as the default unless specified otherwise, and I am not sure if there is some global configuration option to override this. But reinforcing an implicit parameter with an explicit parameter should be no issue for `convert_model.py`, right?

**Comment by Dobiasd** [comment]
The thing is, I guess I will not manage to globally support `channels_first` without increasing the library complexity too much. The export and import of weights already is a mess. It was even way worse back in the days when frugally-deep internally used `channel_first`. Now everything is `channel_last` for at least a bit of simplicity.

But maybe such a global support is not needed for you case, and just the pooling layers have to support this option. Are you interested in contributing this feature?

---

> But reinforcing an implicit parameter with an explicit parameter should be no issue for `convert_model.py`, right?

Sorry, what do you mean by this?

**Comment by hunyadi** [comment]
We definitely do not need global support, the pooling layers would definitely suffice, where the `channels_first` and `channels_last` options (in my opinion) make most sense, and should be relatively straightforward to implement (walk the tensor with a different indexing, aggregating over last dimension or last-but-one dimension).

> Are you interested in contributing this feature?

We have too much on us at the moment, but perhaps in the future? (We will go with the proposed workaround on the short run.)

> Sorry, what do you mean by this?

I meant that it should be no different if I have `channels_last` set explicitly on a layer (e.g. `GlobalAveragePooling1D(data_format="channels_last")`), or if it is implicitly inherited as `channels_last` being the default data layout option in Keras (e.g. `GlobalAveragePooling1D()`). But I think I got the answer from your previous comment.

**Comment by Dobiasd** [comment]
OK, thanks for the explanation.
So if the workaround currently is not a big problem, I guess it's not high priority. Feel free to come back to this issue whenever you'd like to tackle the implementation. :)

**Comment by Dobiasd** [comment]
Closed with [pull request #124 (Add pooling layer parameter "channels_first")](https://github.com/Dobiasd/frugally-deep/pull/124).

Thanks again. :slightly_smiling_face:
